xspec/xspec
# Run end-to-end test for XQuery

This pull request derives from #330. So needs to be handled after that.

---

Some of `[XSpec]/test/end-to-end/cases/*.xspec` files support XQuery. But only one file (`xquery-tutorial.xspec`) is tested.
This pull request enables all XQuery tests by reusing `[XSpec]/test/ant/` in the end-to-end tests.

As a side benefit, this new test scheme runs much faster:

* Travis: Used to process [30 files in 187 seconds](https://travis-ci.org/AirQuick/xspec/jobs/424321159#L1587) vs Now processed [40 files in 44 seconds](https://travis-ci.org/AirQuick/xspec/jobs/424634174#L1587)
* AppVeyor: Used to process [30 files in 197 seconds](https://ci.appveyor.com/project/AirQuick/xspec/build/1197-fix_issue-211_cumulative/job/0k4k8ub8371o42fh#L592) vs Now processed [40 files in 68 seconds](https://ci.appveyor.com/project/AirQuick/xspec/build/1199-e2e-test_ant/job/22gjm9x47a2b4pw9#L592)

### How to run the tests
You can run the end-to-end tests on oXygen or on terminal. See [README](https://github.com/xspec/xspec/tree/2136f43c3ecaefd63f18c3562d776aabef897d64/test/end-to-end) for details.

### Testing this pull request

With Saxon-EE 9.7.0.21, I tested this pull request as follows:

#### Basic
1. Remove all the files in `test\end-to-end\cases\expected\`.
1. Run `test\end-to-end\generate-expected.cmd -silent`.
1. Verify that Git detects no changes.
1. Run `test\end-to-end\run-e2e-tests.cmd -silent`.
1. Verify that the test succeeds.

#### Failure of test result HTML
1. Break the test result HTML intentionally by inserting a dummy text (`<xsl:if test="exists(@query)">_TEST_</xsl:if>`) into [`format-xspec-report.xsl`](https://github.com/xspec/xspec/blob/d774d23313a0034733754b5d0826828fa7ac26fa/src/reporter/format-xspec-report.xsl#L163).
1. Run `run-e2e-tests.cmd`.
1. Verify that the test fails at `xquery-tutorial.xspec`. (Demonstrated by a temporary commit (0757422c41a15e87b299a51e59607265af795b99) on [Travis](https://travis-ci.org/AirQuick/xspec/jobs/424639288#L1604) and [AppVeyor](https://ci.appveyor.com/project/AirQuick/xspec/build/1200-e2e-test_ant/job/bb8cgt1knssqs16p#L618))
1. Run `generate-expected.cmd`.
1. Verify that Git detects changes accordingly.
1. Revert the changes made for the intentional failure.

#### Failure of JUnit report
1. Break the JUnit report intentionally by inserting a dummy text (`<xsl:if test="exists(@query)">_TEST_</xsl:if>`) into [`junit-report.xsl`](https://github.com/xspec/xspec/blob/917654d832b2a4a170b26532620e78754b273647/src/reporter/junit-report.xsl#L33)
1. Run `run-e2e-tests.cmd`.
1. Verify that the test fails at `xquery-tutorial.xspec`.
1. Run `generate-expected.cmd`.
1. Verify that Git detects changes accordingly.
1. Revert the changes made for the intentional failure.

And did the same steps with Saxon-HE 9.8.0.12 using `*.sh` instead of `*.cmd`.

### Further work
Update [Wiki](https://github.com/xspec/xspec/wiki/How-to-Run-the-Test-Suite-Locally#end-to-end-tests)
